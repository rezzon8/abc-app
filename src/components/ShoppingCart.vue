<template>
  <div class="hello">
    <ul>
      <li v-for="(item, index) in products" :key="index">
        <button :class="`add-${item.slug}`" @click="addToCart(item)">
          {{ item.name }} - R{{ item.price }}
        </button>
      </li>
    </ul>
    <ul>
      <template v-if="cart.length > 0">
        <li
          v-for="(item, index) in cart"
          :key="`${item.slug}-${index}`"
          :class="`item-${item.slug}-${index}`"
        >
          {{ item.name }} - R{{ item.price }}
        </li>
      </template>
      <template v-else>
        <li>This cart is empty, why not spoil yourself?</li>
      </template>
    </ul>
    <h2>price: R{{ price }}</h2>
    <h2>tax: R{{ tax }}</h2>
    <h1>total: R{{ total }}</h1>
  </div>
</template>

<script>
export default {
  name: "ShoppingCart",
  props: {
    msg: String
  },
  data() {
    return {
      cart: [],
      products: [
        { name: "Dove Soap", slug: "dove-soap", price: "39.99" },
        { name: "Axe", slug: "axe-deo", price: "99.99" }
      ],
      salesTax: 12.5,
      totalTaxAmount: 0.0
    };
  },
  methods: {
    addToCart(item) {
      return this.cart.push(item);
    }
  },
  computed: {
    price() {
      let total = 0;
      this.cart.forEach(product => {
        total += Number(product.price);
      });
      return total.toFixed(2);
    },
    tax() {
      let total = 0;
      total = Number((this.price / 100) * this.salesTax);
      return total.toFixed(2);
    },
    total() {
      return (Number(this.price) + Number(this.tax)).toFixed(2);
    }
  }
};
</script>

<!-- Add "scoped" attribute to limit CSS to this component only -->
<style scoped lang="scss">
h3 {
  margin: 40px 0 0;
}
ul {
  list-style-type: none;
  padding: 0;
}
li {
  display: inline-block;
  margin: 0 10px;
}
a {
  color: #42b983;
}
</style>
