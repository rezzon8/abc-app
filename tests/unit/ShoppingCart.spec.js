import { shallowMount } from "@vue/test-utils";
import ShoppingCart from "@/components/ShoppingCart.vue";

describe("ShoppingCart.vue", () => {
  it("should render", () => {
    const wrapper = shallowMount(ShoppingCart);
    expect(wrapper.exists()).toBe(true);
    expect(wrapper).toMatchSnapshot();
  });
  it("should render 2 product items(buttons)", () => {
    const wrapper = shallowMount(ShoppingCart);
    expect(wrapper.findAll("button").length).toBe(2);
  });
  it("price, tax and total should be '0.00' to begin", () => {
    const wrapper = shallowMount(ShoppingCart);
    expect(wrapper.vm.price).toBe("0.00");
    expect(wrapper.vm.tax).toBe("0.00");
    expect(wrapper.vm.total).toBe("0.00");
  });
  it("sales tax exist", () => {
    const wrapper = shallowMount(ShoppingCart);
    expect(wrapper.vm.salesTax).toBeDefined();
    expect(wrapper.vm.salesTax).toBe(12.5);
  });
  it("should add 2 Dove products to the cart", () => {
    const wrapper = shallowMount(ShoppingCart);
    wrapper.find("button.add-dove-soap").trigger("click");
    expect(wrapper.vm.price).toBe("39.99");
    expect(wrapper.vm.addToCart).toBeDefined();
    wrapper.find("button.add-dove-soap").trigger("click");
    expect(wrapper.vm.price).toBe("79.98");
  });
  it("should remove an item from the cart", () => {
    const wrapper = shallowMount(ShoppingCart);
    wrapper.find("button.add-dove-soap").trigger("click");
    expect(wrapper.vm.price).toBe("39.99");
    expect(wrapper.vm.removeItem).toBeDefined();
    wrapper.vm.removeItem();
    expect(wrapper.vm.price).toBe("0.00");
  });
  it("should list 5 Dove products, price, tax, total", () => {
    const wrapper = shallowMount(ShoppingCart);
    wrapper.find("button.add-dove-soap").trigger("click");
    wrapper.find("button.add-dove-soap").trigger("click");
    wrapper.find("button.add-dove-soap").trigger("click");
    wrapper.find("button.add-dove-soap").trigger("click");
    wrapper.find("button.add-dove-soap").trigger("click");
    expect(wrapper.vm.cart.length).toEqual(5);
    expect(wrapper.vm.price).toEqual("199.95");
    expect(wrapper.vm.tax).toEqual("24.99");
    expect(wrapper.vm.total).toEqual("224.94");
  });
  it("should list 5 Dove products plus 3 Axe products, price, tax, total", () => {
    const wrapper = shallowMount(ShoppingCart);
    wrapper.find("button.add-dove-soap").trigger("click");
    wrapper.find("button.add-dove-soap").trigger("click");
    wrapper.find("button.add-dove-soap").trigger("click");
    wrapper.find("button.add-dove-soap").trigger("click");
    wrapper.find("button.add-dove-soap").trigger("click");
    expect(wrapper.vm.cart.length).toEqual(5);
    expect(wrapper.vm.price).toBe("199.95");
    expect(wrapper.vm.tax).toEqual("24.99");
    expect(wrapper.vm.total).toEqual("224.94");
    wrapper.find("button.add-axe-deo").trigger("click");
    wrapper.find("button.add-axe-deo").trigger("click");
    wrapper.find("button.add-axe-deo").trigger("click");
    expect(wrapper.vm.cart.length).toEqual(8);
    expect(wrapper.vm.price).toBe("499.92");
    expect(wrapper.vm.tax).toEqual("62.49");
    expect(wrapper.vm.total).toEqual("562.41");
  });
});
