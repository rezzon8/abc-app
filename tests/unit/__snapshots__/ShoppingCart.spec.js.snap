// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ShoppingCart.vue should render 1`] = `
<div class="hello">
  <ul>
    <li><button class="btn waves-effect waves-light input-btn add-dove-soap">
        Dove Soap - R39.99
      </button></li>
    <li><button class="btn waves-effect waves-light input-btn add-axe-deo">
        Axe - R99.99
      </button></li>
  </ul>
  <ul class="cart">
    <li>This cart is empty, why not spoil yourself?</li>
  </ul>
  <h2>price: R0.00</h2>
  <h2>tax: R0.00</h2>
  <h1>total: R0.00</h1>
</div>
`;
